### PR TITLE
fix Issue 23089 - Linkage-related ICE regression in v2.100.0-rc.1

### PR DESCRIPTION
**Pull request by ibuclaw**
Exactly the same change to tocsym.d as in #13942, but applied to VarDeclaration symbols.

### CODE REVIEW

**Review by ljmf00 — APPROVED** [review]

File: test/compilable/test23089.d [hunk 1]
```diff
@@ -0,0 +1,7 @@
+// https://issues.dlang.org/show_bug.cgi?id=23089
```
Comment (ljmf00) [inline test/compilable/test23089.d:1]: You can use the trait getLinkage here to verify the applied linkage.

**Review by dkorpel — APPROVED** [review]

**Comment by dkorpel** [comment]
Thanks for fixing it this quickly. I thought the case was covered by [test/compilable/json.d](https://github.com/dlang/dmd/pull/13942/files#diff-43fb2e1e984c1174584fbe0820036731c67e359f51cfdf05764386cecb2c7959), but that test doesn't do code gen. 

It still trips an assert on macOS for some reason though.

**Comment by kinke** [comment]
> Exactly the same change to tocsym.d as in https://github.com/dlang/dmd/pull/13942

Hmm, so the glue layer is newly supposed to handle `extern(System)`? LDC doesn't so far; not too fond of that requirement. I guess we could also hide the explicit one as `_linkage` field, and expose a `linkage()` getter that takes care of resolving the `system` and `default` ones.

**Comment by kinke** [comment]
E.g., this looks IMO dangerous too: https://github.com/dlang/dmd/blob/3a66d2fb00008924215f40b0cc19879a2ea1c80e/src/dmd/cppmangle.d#L1879-L1880

**Comment by kinke** [comment]
FWIW, the LDC glue layer has some 11 usages of `Declaration::linkage` which would basically all need further resolving.

**Comment by kinke** [comment]
FWIW, the LDC glue layer has some 11 usages of `Declaration::linkage` which would basically all need further resolving.

**Comment by ibuclaw** [comment]
> > Exactly the same change to tocsym.d as in #13942
> 
> Hmm, so the glue layer is newly supposed to handle `extern(System)`? LDC doesn't so far; not too fond of that requirement. I guess we could also hide the explicit one as `_linkage` field, and expose a `linkage()` getter that takes care of resolving the `system` and `default` ones.

I pretty much ignore what the value of `linkage` is for variables, and it is resolved by requesting the mangling IIUC @dkorpel's original PR correctly.  Ditto function declarations too, except for the generation of its type - which is eagerly resolved by TypeFunction semantic.

**Comment by kinke** [comment]
I guess I could replace the `FuncDeclaration::linkage` usages by `fd->type->toBasetype()->isTypeFunction()->linkage`, but meh, ugly.

Considering https://github.com/dlang/dmd/pull/13942 introduced 2 regressions already, and just makes the getLinkage trait return unresolved `System`, I'm more inclined to reverting it for the 2.100 release - I'm afraid of more regressions appearing with that hardly used linkage.

**Comment by kinke** [comment]
Wrt. mangling - I don't see where the frontend would actually handle Win32's stdcall mangling anywhere. With LDC and `-mtriple=i686-windows-msvc`:
```D
extern (Windows) int foo(int, bool) { return 0; } // mangled: _foo@8
pragma(msg, foo.mangleof); // "foo"
```

**Comment by ibuclaw** [comment]
> Wrt. mangling - I don't see where the frontend would actually handle Win32's stdcall mangling anywhere. With LDC and `-mtriple=i686-windows-msvc`:

It doesn't, any system-ABI mangling is taken care of by the backend.

There was a reason why I removed the front-end hack of adding double underscores to OSX mangling.

File: test/compilable/test23089.d (hunk 1 above)
Comment (ibuclaw) [reply test/compilable/test23089.d:1]: It's a backend ICE, not a front-end.

**Comment by kinke** [comment]
> It doesn't, any system-ABI mangling is taken care of by the backend.
> 
> There was a reason why I removed the front-end hack of removing the leading underscore to OSX mangling.

I know this is pretty off-topic, but I definitely wouldn't consider stdcall to be a "system ABI", contrary to a platform-specific C mangling prefix like Mac's underscore prefix. Win32 is hopefully dying for good, which will make both `extern(Windows)` and `extern(System)` useless legacy stuff, together with this `.mangleof` 'lie' (for all other linkages, you get the platform-agnostic 'C' mangle compatible with `pragma(mangle)`).

**Review by kinke — APPROVED** [review]
Anyway, this is a fix for the status quo.

**Pull request merged**